flent/runners.py

```diff
@@ -950,55 +959,180 @@ def find_binary(cls, host, interval, length, ip_version, local_bind=None,
+            pid = self.pid_re.search(part)
+            if None is pid:
+                continue
+            pid_str = pid.group('pid')
+
+            f_ports = self.ports_ipv4_re.search(sp)
+            if None is f_ports:
+                f_ports = self.ports_ipv6_re.search(sp)
+
+            if None is f_ports:
+                raise ParseException()
+
+            dst_p = str(f_ports.group('dst_p').split(":")[-1])
+
+            if self.pid == pid_str and dst_p != self.ctrl_port:
+                sub_part.append(sp)
+
+        if 1 != len(sub_part):
+            raise ParseException()
+
+        return sub_part[0]
+
+    def parse_part(self, part):
+
+        sub_part = self.filter_np_parent(part)
+
+        timestamp = self.time_re.search(sub_part)
+        cwnd = self.cwnd_re.search(sub_part)
+        rtt = self.rtt_re.search(sub_part)
+
+        if None in [timestamp, cwnd, rtt]:
+            raise ParseException()
+
+        timestamp = float(timestamp.group('timestamp'))
+        cwnd = int(cwnd.group('cwnd'))
+        rtt, rtt_var = [float(x) for x in rtt.group('rtt').split('/')]
+
+        raw_values = {'t': timestamp, 'cwnd': cwnd, 'rtt': rtt, 'rtt_var': rtt_var}
+        self.raw_values.append(raw_values)
+
+        return raw_values
+
+    def parse(self, output):
+        results = {}
+        parts = output.split("\n---\n")
+        for part in parts:
+            try:
+                res_dict = self.parse_part(part)
+                t = res_dict['t']
+                for k, v in res_dict.items():
+                    if k == 't':
+                        continue
+                    if k not in results:
+                        results[k] = [[t, v]]
+                    else:
+                        results[k].append([t, v])
+
+            except ParseException:
+                continue
+
+        return results
+
+    def find_binary(self, ip_version, host, interval, length, target):
+        script = os.path.join(DATA_DIR, 'scripts', 'ss_iterate.sh')
+        if not os.path.exists(script):
+            raise RuntimeError("Cannot find ss_iterate.sh.")
+
+        bash = util.which('bash')
+        if not bash:
+            raise RuntimeError("Socket stats requires a Bash shell.")
+
+        resol_target = util.lookup_host(target, ip_version)[4][0]
+        if ip_version == 6:
+            resol_target = "[" + str(resol_target) + "]"
+
+        return "{bash} {script} -I {interval:.2f} " \
+                "-c {count:.0f} -H {host} -t '{target}'".format(
+                        bash=bash,
+                        script=script,
+                        interval=interval,
+                        count=length // interval + 1,
+                        host=host,
+                        target=resol_target)
+
+
 class TcRunner(ProcessRunner):
-    """Runner for iterated `tc -s qdisc`. Expects iterations to be separated by
+                    """Runner for iterated `tc -s qdisc`. Expects iterations to be separated by
```

Comment:
Vim key combination accident ... cured it. Need to praise **git add -p** once more.